thoth/lab/adviser.py

```diff
@@ -0,0 +1,322 @@
+        adviser_data, index=["datetime", "analyzer_version", "error", "justification", "message", "type"]
+    )
+    adviser_df = adviser_df.transpose()
+    adviser_df["date"] = pd.to_datetime(adviser_df["datetime"])
+
+    return adviser_df
+
+
+def extract_adviser_justifications(report: Dict[str, Any], adviser_dict: Dict[str, Any], ids: str) -> Dict[str, Any]:
+    """Retrieve justifications from adviser results."""
+    if not report:
+        return adviser_dict
+
+    products = report.get("products")
+    adviser_dict = extract_justifications_from_products(products=products, adviser_dict=adviser_dict, ids=ids)
+
+    return adviser_dict
+
+
+def extract_justifications_from_products(
+    products: List[Dict[str, Any]], adviser_dict: Dict[str, Any], ids: str
+) -> Dict[str, Any]:
+    """Extract justifications from products in adviser results."""
+    if not products:
+        return adviser_dict
+
+    for product in products:
+        justification = product["justification"]
+        if justification:
+            # TODO: Consider all justifications
+            adviser_dict[ids] = {}
+            adviser_dict[ids]["justification"] = justification
+            adviser_dict[ids]["error"] = False
+            adviser_dict[ids]["message"] = justification[0]["message"]
+            adviser_dict[ids]["type"] = justification[0]["type"]
+
+    return adviser_dict
+
+
+def create_final_dataframe(adviser_dataframe: pd.DataFrame) -> pd.DataFrame:
+    """Create final dataframe with all information required for plots.
+
+    :param adviser_dataframe: data frame as returned by `aggregate_adviser_results` method.
+    """
+    jm_encoding = []
+    for index, row in adviser_dataframe[["message"]].iterrows():
+        hash_object = hashlib.sha256(bytes(row.values[0], "raw_unicode_escape"))
+        hex_dig = hash_object.hexdigest()
+        jm_encoding.append([index, row.values, hex_dig])
+
+    label_encoder = LabelEncoder()
+    justification_result = copy.deepcopy(adviser_dataframe.to_dict())
+
+    jm_hash_id_values = array([pp[2] for pp in jm_encoding])
+    integer_jm_hash_id_values_encoded = label_encoder.fit_transform(jm_hash_id_values)
+
+    counter = 0
+    for id_jm in integer_jm_hash_id_values_encoded:
+        jm_encoding[counter] = jm_encoding[counter] + [id_jm]
+        counter += 1
+
+    justification_result["jm_hash_id_encoded"] = {el[0]: el[3] for el in jm_encoding}
+
+    final_dataframe = pd.DataFrame(justification_result)
+
+    return final_dataframe
+
+
+def create_adviser_results_histogram(plot_df: pd.DataFrame):
+    """Create inspection performance parameters plot in 3D.
+
+    :param plot_df dataframe for plot of adviser results
+    """
+    histogram_data = {}
+
+    for index, row in plot_df[["jm_hash_id_encoded", "message", "type"]].iterrows():
+        encoded_id = row["jm_hash_id_encoded"]
+        if row["jm_hash_id_encoded"] not in histogram_data.keys():
+            histogram_data[encoded_id] = {}
```

Comment:
This can be done in a single assignment of the resulting dict. ( just a nitpick :) )

Reply:
I like it :) Thanks I will change it :)